File: Labos 1/lab1.py
MAX_N = 15000

def polinomi(a, b, size_a, size_b):
    result = [0] * (MAX_N*2 + 1)
    for i in range(min(size_a, MAX_N)):
        for j in range(min(size_b, MAX_N)):
            result[i + j] += a[i] * b[j]
    return result

def sol(a, b, c, d, e, n):
    a = n if (a == -1 or a > n) else a
    b = n if (b == -1 or b > n) else b
    c = n if (c == -1 or c > n) else c
    d = n if (d == -1 or d > n) else d
    e = n if (e == -1 or e > n) else e

    A = [1] * (min(a + 1, MAX_N + 1))
    B = [1] * (min(b + 1, MAX_N + 1))
    C = [1] * (min(c + 1, MAX_N + 1))
    D = [1] * (min(d + 1, MAX_N + 1))
    E = [1] * (min(e + 1, MAX_N + 1))

    temp = polinomi(A, B, a + 1, b + 1)
    R = polinomi(temp, C, a + b + 1, c + 1)
    temp = polinomi(R, D, a + b + c + 1, d + 1)
    R = polinomi(temp, E, a + b + c + d + 1, e + 1)

    return R[n]

def main():
    print("Unesite broj vrecica caja Bohea (parameter a):", end=" ")
    a = int(input())
    print("Unesite broj vrecica caja Congo  (parameter b):", end=" ")
    b = int(input())
    print("Unesite broj vrecica caja Souch gs (parameter c):", end=" ")
    c = int(input())
    print("Unesite broj vrecica caja Singl  (parameter d):", end=" ")
    d = int(input())
    print("Unesite broj vrecica caja Hyson (parameter e):", end=" ")
    e = int(input())
    print("Unesite broj n:", end=" ")
    n = int(input())

    result = sol(a, b, c, d, e, n)
    print(f"Gradjani bostona mogu napraviti {result} razlicitih cajanki.")

if __name__ == "__main__":
    main()

File: Labos 1/lab1.cxx
#include <cstdio>
#include <iostream>

typedef long long ll;

#define pb push_back

using namespace std;

string cajevi[5] = {"Bohea", "Congou", "Souchong", "Singlo", "Hyso"};
char parametar[5] = {'a', 'b', 'c', 'd', 'e'};
int broj_cajeva[5], n;

const int MAX_N = 15000;

ll dp[15001];

void input() {
  // test case iz primjera
  // freopen("test.txt", "r", stdin);

  for (int i = 0; i < 5; i++) {
    cout << "Unesite broj vrecica caja " << cajevi[i] << " (parametar "
         << parametar[i] << "): ";
    cin >> broj_cajeva[i];
  }

  cout << "Unesite broj n: ";
  cin >> n;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);

  input();

  dp[0] = 1;

  for (int caj : broj_cajeva) {
    for (int i = n; i >= 0; --i) {
      int max = (caj == -1 || caj > n) ? n - i : caj;
      for (int j = 1; j <= max && i + j <= n; ++j) {
        dp[i + j] += dp[i];
      }
    }
  }

  //   for (int i = 0; i <= n; i++) {
  // cout << "dp[" << i << "] = " << dp[i] << endl;
  //   }

  //   cout << dp[n];

  cout << "Gradjani Bostona mogu napraviti " << dp[n] << " razlicitih cajanki.";

  return 0;
}